specs/core/0_beacon-chain.md

```diff
@@ -1442,39 +1442,44 @@ def get_changed_validators(validators: List[ValidatorRecord],
     """
     # The active validators
     active_validator_indices = get_active_validator_indices(validators)
-    # The total balance of active validators
+    # The total effective balance of active validators
     total_balance = sum([get_effective_balance(v) for i, v in enumerate(validators) if i in active_validator_indices])
-    # The maximum total balance in Gwei that can be deposited or withdrawn
-    max_allowable_balance_changes = max(
+    
+    # The maximum balance churn in Gwei (for deposits and exits separately)
+    max_balance_churn = max(
         MAX_DEPOSIT * GWEI_PER_ETH,
         total_balance // (2 * MAX_BALANCE_CHURN_QUOTIENT)
     )
-    # Go through the list start to end, depositing and withdrawing as many as possible
+    
+    # Activate validators within the allowable balance churn
     total_activated_balance = 0
-    total_exited_balance = 0
     for i in range(len(validators)):
         if validators[i].status == PENDING_ACTIVATION:
             validators[i].status = ACTIVE
             total_activated_balance += get_effective_balance(validators[i])
+            if total_activated_balance >= max_balance_churn:
```

Comment:
Why? If we go above `max_balance_churn` should we avoid activating that validator?

Reply:
But you already set the validator to `ACTIVE` in the previous line.

You can check if activating the validator _would_ push the `total_activated_balance` over the edge and if not, do the activation, else break.

One reason to allow the activation/exit of the validator that _does_ throw the number over the `max_balance_churn` is to ensure that new validators can register in times of extremely low participation.  If there were only ~30 validators, and you did the check before activation, then no one would ever be able to register again.

Reply:
> But you already set the validator to `ACTIVE` in the previous line.

Oh right, fixed that.

> If there were only ~30 validators, and you did the check before activation, then no one would ever be able to register again.

Does the `max(MAX_DEPOSIT * GWEI_PER_ETH, ` logic prevent that?

Reply:
ah, you're right. Because a validator can top-up their balance before they are `ACTIVE`, you could get into a situation where the 1st `PENDING_ACTIVATION` validator in the list has a balance of `33 ETH` which is higher than `max_balance_churn` -- thus halting registration forever. (this is an extremely unlikely event)

One more nitpick is that `total_activated_balance` is no longer descriptive as it captures the value of the last validator that _was not_ activated. This is fine because we don't do anything else with this var, but I could see it causing confusion during debugging.